### PR TITLE
preview_site_jekyll: Fix breakage due to Ubuntu-installed jekyll

### PR DESCRIPTION
**Pull request by EricCousineau-TRI**
Proactively pin versions

Error on current `gh-pages` (d61cc1d06dbe994464c61c95e455c4b362c536fb):

```
$ git clean -fxd
$ ./preview_site_jekyll.py
...
{base}/styleguide/_gems/gems/jekyll-3.9.0/lib/jekyll/drops/document_drop.rb:8: warning: already initialized constant Jekyll::Drops::DocumentDrop::NESTED_OBJECT_FIELD_BLACKLIST
/usr/lib/ruby/vendor_ruby/jekyll/drops/document_drop.rb:8: warning: previous definition of NESTED_OBJECT_FIELD_BLACKLIST was here
{base}/styleguide/_gems/gems/jekyll-3.9.0/lib/jekyll/drops/drop.rb:8: warning: already initialized constant Jekyll::Drops::Drop::NON_CONTENT_METHODS
/usr/lib/ruby/vendor_ruby/jekyll/drops/drop.rb:8: warning: previous definition of NON_CONTENT_METHODS was here
/usr/lib/ruby/vendor_ruby/jekyll/hooks.rb:56: warning: constant ::Fixnum is deprecated
Configuration file: {base}/styleguide/_config.yml
Configuration file: {base}/styleguide/_config.yml
Configuration file: {base}/styleguide/_config.yml
jekyll 3.1.6 | Error:  undefined method `config=' for #<Jekyll::Site:0x00005637e7de1bb8>
Did you mean?  config
Traceback (most recent call last):
  File "./preview_site_jekyll.py", line 92, in <module>
    main()
  File "./preview_site_jekyll.py", line 88, in main
    run_env(["bundle", "exec", "jekyll", "serve", "--no-watch"])
  File "./preview_site_jekyll.py", line 80, in run_env
    run(args, env=env, check=True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['bundle', 'exec', 'jekyll', 'serve', '--no-watch']' returned non-zero exit status 1.
```

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)

### CODE REVIEW

**Review by EricCousineau-TRI — COMMENTED** [review]
+@jwnimmer-tri for review, please.

\cc @jamiesnape @betsymcphail 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-:-MTf2wE0DnAKY8Dc6h2N:b-xk8852)* status: LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Comment by jamiesnape** [comment]
<s>If we are going to be using gems, there is a lot we could do differently, especially plugin-wise (https://rubygems.org/gems/jekyll-relative-links is a must)</s>. I don't think your comments are right about bundler, though. I use it frequently and it is completely isolated. Also rbenv covers both gems and ruby versions, but the default ruby version is system (you are confusing ruby-build, I think).

**Comment by jamiesnape** [comment]
rvm is scary, though. That's why rbenv exists.

**Comment by EricCousineau-TRI** [comment]
The issue with `bundler` is that I don't know how to install it without polluting something.

Debian's version of `bundler` is too old to support the `Gemfile` that's there, and I don't feel like to trying to make things work with the older `bundler`.

If that were fixed, then yeah, I'd be down. But I still want to isolate `bundler` itself.

**Comment by jamiesnape** [comment]
Then probably you would have to go `rbenv`, with or without `bundler` (with is nicer), but I see what you mean.

**Comment by jamiesnape** [comment]
Not obvious what is too advanced in the `Gemfile`, they have not changed the format too much over time, but if bundler is not happy it must be something.

**Comment by EricCousineau-TRI** [comment]
My naive reading of `rbenv` is that I still have to futz with environment variables to isolate my changes from the user (e.g. interpreter clash). I don't see anything about setting up an explicit root, aside from `RBENV_ROOT`.

I'll modify my docs, but I don't see that usage as anything better. (Unless it solves the `--local` problem, which I don't want to test right now.)

**Comment by jamiesnape** [comment]
I'm not saying change, just what my. natural workflow is. Keep everything as is.

**Comment by jamiesnape** [comment]
`rbenv` is nice enough that someone wrote a Python equivalent (https://github.com/pyenv/pyenv), so I do certainly recommend it, if you were ever to do significant Ruby development.

**Comment by jwnimmer-tri** [comment]
I'll take a look at this later this week on my Drake day, but also just to clarify the endgame here in my head...

I want to move the integration of the cppguide into our website from this repository's GH pages into the `@drake//doc:build` tool, such that we have a unified preview & deploy engine inside Drake itself.  (Our WORKSPACE already fetches the styleguide repository for cpplint tooling; it will also then use the documentation contents in this new world.) Perhaps this preview tooling just moves into drake/doc as part of that (as a subroutine), or perhaps Drake's existing jekyll deploy gains this one markdown file to eat as well.

So in short, it's possible that this preview tool has a short shelf live, so I'm not going to worry yet about polishing it too much, instead just confirming that it works.

**Comment by EricCousineau-TRI** [comment]
Sounds great to me! When you do so, can you also confirm it still fails on your machine on `gh-pages` if you have `jekyll` installed via `apt`?

**Review by jwnimmer-tri — COMMENTED** [review]
Confirmed that (without using this PR's fixes) the preview script breaking the same way as Eric posted in this PR description.

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-:-MU9UmVT9mpOpepyJkei:b-iuwb6p)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MU9VZDR1Yke-afx-uG9:-MU9VZDSCq7qTngedCM9:b-n9wtm6) (no related file):*
As of this PR's 9152bbef8938e7fc03e4df43b26f26be5de12663, I am seeing the following error:

```console
jwnimmer@call-cps:~/jwnimmer-tri/styleguide$ ./preview_site_jekyll.py 
Setting GEM_HOME to tmpdir: /home/jwnimmer/jwnimmer-tri/styleguide/_gems
Setting HOME to /home/jwnimmer/jwnimmer-tri/styleguide/_gems/fake_home


Setting up site. May take a while the first time...

+ bash -c 'export -p'
declare -x GEM_HOME="/home/jwnimmer/jwnimmer-tri/styleguide/_gems"
declare -x HOME="/home/jwnimmer/jwnimmer-tri/styleguide/_gems/fake_home"
declare -x OLDPWD
declare -x PATH="/home/jwnimmer/jwnimmer-tri/styleguide/_gems/bin:/usr/bin:/bin"
declare -x PWD="/home/jwnimmer/jwnimmer-tri/styleguide"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x TERM="xterm-256color"
declare -x USER="jwnimmer"
+ gem install -v '~> 2.2' bundler
Fetching: bundler-2.2.11.gem (100%)
Successfully installed bundler-2.2.11
Parsing documentation for bundler-2.2.11
Installing ri documentation for bundler-2.2.11
Done installing documentation for bundler after 3 seconds
1 gem installed
+ bundle config set --local path /home/jwnimmer/jwnimmer-tri/styleguide/_gems
+ bundle install
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Bundler could not find compatible versions for gem "public_suffix":
  In snapshot (Gemfile.lock):
    public_suffix (= 4.0.6)

  In Gemfile:
    jekyll (~> 3.9) was resolved to 3.9.0, which depends on
      addressable (~> 2.4) was resolved to 2.7.0, which depends on
        public_suffix (< 5.0, >= 2.0.2)

    github-pages (= 211) was resolved to 211, which depends on
      github-pages-health-check (= 1.16.1) was resolved to 1.16.1, which depends on
        public_suffix (~> 3.0)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
Traceback (most recent call last):
  File "./preview_site_jekyll.py", line 136, in <module>
    main()
  File "./preview_site_jekyll.py", line 128, in main
    run_env(["bundle", "install"])
  File "./preview_site_jekyll.py", line 115, in run_env
    run(args, env=env, check=True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['bundle', 'install']' returned non-zero exit status 6.
```

As a fix, I tried `rm -rf Gemfile _gems /tmp/robotlocomotion-styleguide-jekyll-home` first and then re-ran.  No dice.

---

**Comment by jamiesnape** [comment]
You should just need only the following in the `Gemfile` and nothing else, and if you use `bundle`, you should be fine.

```ruby
source 'https://rubygems.org'
gem 'github-pages', '~> 212'
```

I need to look at what we are doing that is unusual.

**Comment by jamiesnape** [comment]
(also `echo 'gem: --no-document' > ~/.gemrc` will save you a load of time)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MU9VZDR1Yke-afx-uG9:-MUA5pxA3LigGZkZrDl1:b3y4xlw) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

As of this PR's 9152bbef8938e7fc03e4df43b26f26be5de12663, I am seeing the following error:

```console
jwnimmer@call-cps:~/jwnimmer-tri/styleguide$ ./preview_site_jekyll.py 
Setting GEM_HOME to tmpdir: /home/jwnimmer/jwnimmer-tri/styleguide/_gems
Setting HOME to /home/jwnimmer/jwnimmer-tri/styleguide/_gems/fake_home


Setting up site. May take a while the first time...

+ bash -c 'export -p'
declare -x GEM_HOME="/home/jwnimmer/jwnimmer-tri/styleguide/_gems"
declare -x HOME="/home/jwnimmer/jwnimmer-tri/styleguide/_gems/fake_home"
declare -x OLDPWD
declare -x PATH="/home/jwnimmer/jwnimmer-tri/styleguide/_gems/bin:/usr/bin:/bin"
declare -x PWD="/home/jwnimmer/jwnimmer-tri/styleguide"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x TERM="xterm-256color"
declare -x USER="jwnimmer"
+ gem install -v '~> 2.2' bundler
Fetching: bundler-2.2.11.gem (100%)
Successfully installed bundler-2.2.11
Parsing documentation for bundler-2.2.11
Installing ri documentation for bundler-2.2.11
Done installing documentation for bundler after 3 seconds
1 gem installed
+ bundle config set --local path /home/jwnimmer/jwnimmer-tri/styleguide/_gems
+ bundle install
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Bundler could not find compatible versions for gem "public_suffix":
  In snapshot (Gemfile.lock):
    public_suffix (= 4.0.6)

  In Gemfile:
    jekyll (~> 3.9) was resolved to 3.9.0, which depends on
      addressable (~> 2.4) was resolved to 2.7.0, which depends on
        public_suffix (< 5.0, >= 2.0.2)

    github-pages (= 211) was resolved to 211, which depends on
      github-pages-health-check (= 1.16.1) was resolved to 1.16.1, which depends on
        public_suffix (~> 3.0)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
Traceback (most recent call last):
  File "./preview_site_jekyll.py", line 136, in <module>
    main()
  File "./preview_site_jekyll.py", line 128, in main
    run_env(["bundle", "install"])
  File "./preview_site_jekyll.py", line 115, in run_env
    run(args, env=env, check=True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['bundle', 'install']' returned non-zero exit status 6.
```

As a fix, I tried `rm -rf Gemfile _gems /tmp/robotlocomotion-styleguide-jekyll-home` first and then re-ran.  No dice.
</blockquote></details>

Confirmed fixed in latest push.

---
*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MUA5hLp20mfBwOOZltG:-MUA5hLq2hR7vrvau1j8:bee230n) (no related file):*
TRI computers have management software listening on port 4000, so Jekyll's default preview at 4000 was erroring out for me.  Here was my fix:

```patch
jwnimmer@call-cps:~/jwnimmer-tri/styleguide$ git diff
diff --git a/preview_site_jekyll.py b/preview_site_jekyll.py
index 6ab472c..bd21c41 100755
--- a/preview_site_jekyll.py
+++ b/preview_site_jekyll.py
@@ -135,7 +135,7 @@ def main():
 
     # Use --no-watch because using the gem tmpdir under this folder may
     # overload inotify.
-    run_env(["bundle", "exec", "jekyll", "serve", "--no-watch"])
+    run_env(["bundle", "exec", "jekyll", "serve", "--no-watch", "--port", "8000"])
 
 
 assert __name__ == "__main__"
```

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MUA5mVL4mdAPQ6fNA7R:-MUA5mVL4mdAPQ6fNA7S:b-tkvh2y) (no related file):*
Working: Realized that with these changes, I can't see the generated `pyguide.html` file - it only has the raw `pyguide.md` file.

Will try for 2min more to fix.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)* status: 2 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MUA5hLp20mfBwOOZltG:-MUA6ChXDBsSRMijUWk_:b-qsx7uj) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

TRI computers have management software listening on port 4000, so Jekyll's default preview at 4000 was erroring out for me.  Here was my fix:

```patch
jwnimmer@call-cps:~/jwnimmer-tri/styleguide$ git diff
diff --git a/preview_site_jekyll.py b/preview_site_jekyll.py
index 6ab472c..bd21c41 100755
--- a/preview_site_jekyll.py
+++ b/preview_site_jekyll.py
@@ -135,7 +135,7 @@ def main():
 
     # Use --no-watch because using the gem tmpdir under this folder may
     # overload inotify.
-    run_env(["bundle", "exec", "jekyll", "serve", "--no-watch"])
+    run_env(["bundle", "exec", "jekyll", "serve", "--no-watch", "--port", "8000"])
 
 
 assert __name__ == "__main__"
```
</blockquote></details>

Done. Oops!

---

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MUA5mVL4mdAPQ6fNA7R:-MUA6JrjDbE1h3QAvzKZ:b-uiyl6e) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Working: Realized that with these changes, I can't see the generated `pyguide.html` file - it only has the raw `pyguide.md` file.

Will try for 2min more to fix.
</blockquote></details>

I had the same experience.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/styleguide/42)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/styleguide/42#-MUA5mVL4mdAPQ6fNA7R:-MUABpQe5XNcqIe7xzjT:b-ckmrcq) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I had the same experience.
</blockquote></details>

Aye. Tried diff. versions, solution is non-obvious. Punting for now (will let PR grow stale).

---

**Comment by jamiesnape** [comment]
Just push changes to your fork and let GitHub do the work.

(or buy a Mac)

**Comment by jwnimmer-tri** [comment]
FYI I have a different solution ready to PR to Drake soon.  I have a change to `@drake//doc:build` which rolls the styleguide publishing into the main deployment process, without using the styleguide's GitHub pages under hood the anymore.  So I'd advise holding off on any more work here until we see if that pans out.

_Edit: Posted at https://github.com/RobotLocomotion/drake/pull/14811 now_.

**Comment by EricCousineau-TRI** [comment]
Oooh, that'd be awesome!

> Just push changes to your fork and let GitHub do the work.

Aye, suboptimal, but works.

> (or buy a Mac)

wayyyyyy suboptimal for me :P

Closing for now!
